### PR TITLE
Add model saving and loading examples

### LINKED ISSUE #1156: Burn Book: Missing recorder examples
We should add more comprehensive examples and documentation on how to load and save weights.

### PR DESCRIPTION
**Pull request by laggui**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues

Closes #1156

### Changes

- Added precision settings table in the Record section of the book
- Added Saving & Loading Models section with examples
  - Using `save_file()` and `load_file()`
  - Using `init_with(record)` to directly initialize the module with the record weights

If I am missing use cases let me know! I'll add them based on the requested changes.

### CODE REVIEW

File: burn-book/src/saving-and-loading.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+```rust, ignore
+// Save model in MessagePack format with full precision
+let recorder = NamedMpkFileRecorder::<FullPrecisionSettings>::new();
+model
+  .save_file(model_path, &recorder)
+  .expect("Could not save model");
+```
+
+Note that the file extension is automatically handled by the recorder depending on the one you choose. Therefore, only the file path and base name can be provided.
```
Comment (nathanielsimard) [inline burn-book/src/saving-and-loading.md:13]: can be provided => should be provided.

File: burn-book/src/saving-and-loading.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
+# Saving and Loading Models
+
+Saving your trained machine learning model to disk is quite easy, no matter the output format you choose. As mentioned in the [Record](./building-blocks/record.md) section, different formats are supported to serialize/deserialize models. By default, we use the `NamedMpkFileRecorder` which uses the [MessagePack](https://msgpack.org/) binary serialization format with the help of [smp_serde](https://docs.rs/rmp-serde/).
+
+```rust, ignore
+// Save model in MessagePack format with full precision
+let recorder = NamedMpkFileRecorder::<FullPrecisionSettings>::new();
+model
+  .save_file(model_path, &recorder)
+  .expect("Could not save model");
```
Comment (nathanielsimard) [inline burn-book/src/saving-and-loading.md:10]: "Could not save the model" => "Should be able to save the model".

File: burn-book/src/saving-and-loading.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+Note that the file extension is automatically handled by the recorder depending on the one you choose. Therefore, only the file path and base name can be provided.
+
+Now that you have a trained model saved to your disk, you can easily load it in a similar fashion.
+
+```rust, ignore
+// Load model in full precision from MessagePack file
+let recorder = NamedMpkFileRecorder::<FullPrecisionSettings>::new();
+model
+  .load_file(model_path, &recorder)
+  .expect("Failed to load model file");
```
Comment (nathanielsimard) [inline burn-book/src/saving-and-loading.md:22]: Should be able to load the model's weights from the provided file.

File: burn-book/src/saving-and-loading.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+}
+```
+
+Now, let's save a model that we can load later. In the following snippets, we use `type Backend = NdArray<f32>` but you can use whatever backend you like.
+
+```rust, ignore
+// Create a dummy initialized model to save
+let model = Model::<Backend>::new(&Default::default());
```
Comment (nathanielsimard) [inline burn-book/src/saving-and-loading.md:70]: I would add an extra line `let device = Default::default();` just to be more specific.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Very good documentation, some minot comments but nothing major. We will need to update a bit the text after this PR is merged: #1165

File: burn-book/src/saving-and-loading.md (hunk 2 above)
Comment (laggui) [reply burn-book/src/saving-and-loading.md]: Oops, still not familiar with the conventions for this 😄

**Review by nathanielsimard — APPROVED** [review]

File: burn-book/src/saving-and-loading.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// Load model in full precision from MessagePack file
+let recorder = NamedMpkFileRecorder::<FullPrecisionSettings>::new();
+model
+  .load_file(model_path, &recorder)
+  .expect("Should be able to load the model weights from the provided file");
+```
+
+**Note:** models can be saved in different output formats, just make sure you are using the correct recorder type when loading the saved model. Type conversion between different precision settings is automatically handled, but formats are not interchangeable. A model can be loaded from one format and saved to another format, just as long as you load it back with the new recorder type afterwards.
+
+## Custom Initialization from Recorded Weights
```
Comment (antimora) [inline burn-book/src/saving-and-loading.md:27]: I would remove "Custom" because it applies there is an alternative normal path, which isn't the case.

**Review by antimora — CHANGES_REQUESTED** [review]
Overall looks good, however, here is what I have add:
1. Text needs to be formatted using `prettier` (we have a `.prettierrc.json` configuration checked) to wrap text around properly ( make the length = 100).
2. We also need to add text on how to store serialized data in the bytes as part of binary executable. This example explains: https://github.com/tracel-ai/burn/blob/main/examples/mnist-inference-web/src/state.rs. This is a super important feature for no-std and WASM build, since they may not have access to filesystem.
3. Remove Custom word from the "Custom Initialization from Recorded Weights" section.

File: burn-book/src/saving-and-loading.md (hunk 5 above)
Comment (laggui) [reply burn-book/src/saving-and-loading.md:27]: Good point! 👍

File: burn-book/src/saving-and-loading.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
+# Saving and Loading Models
+
+Saving your trained machine learning model to disk is quite easy, no matter the output format you choose. As mentioned in the [Record](./building-blocks/record.md) section, different formats are supported to serialize/deserialize models. By default, we use the `NamedMpkFileRecorder` which uses the [MessagePack](https://msgpack.org/) binary serialization format with the help of [smp_serde](https://docs.rs/rmp-serde/).
```
[the remark this answers is not in the document]
Comment (laggui) [reply burn-book/src/saving-and-loading.md:3]: To remove ambiguity I guess we could simply say "Saving your trained machine learning model is quite easy, no matter the output format you choose". Lmk :)

**Comment by laggui** [comment]
> Overall looks good, however, here is what I have add:
> 
>     1. Text needs to be formatted using `prettier` (we have a `.prettierrc.json` configuration checked) to wrap text around properly ( make the length = 100).
> 
>     2. We also need to add text on how to store serialized data in the bytes as part of binary executable. This example explains: https://github.com/tracel-ai/burn/blob/main/examples/mnist-inference-web/src/state.rs. This is a super important feature for no-std and WASM build, since they may not have access to filesystem.
> 
>     3. Remove Custom word from the "Custom Initialization from Recorded Weights" section.

Was not aware of `prettier` for formatting, thank you!

Will make the requested changes first thing tomorrow. Thanks a lot for the feedback!

**Comment by laggui** [comment]
For the `BinByteRecorder` use-cases there were no indications regarding how the model was saved, the existing examples only loaded an already saved model. I think it was still pretty straightforward but let me know if there are other ways to go about saving + loading (including) the model.

**Review by antimora — APPROVED** [review]
Looks great. Thank you for your improvements.

**Pull request merged**